staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,131 @@
+
+	// Items is the list of groups for discovery.
+	Items []APIGroupDiscovery `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscovery holds information about which resources are being served for all version of the API Group.
+type APIGroupDiscovery struct {
+	TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// The only field completed will be Name.  For instance, resourceVersion will be empty.
+	// Name is the name of the API group whose discovery information is presented here.
+	// Name is allowed to be "" to represent the legacy, ungroupified resources.
+	// This is included to make the standard deserialization easier.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Versions are the versions supported in this group. They are sorted in descending order of preference,
+	// with the preferred version being the first entry.
+	// +listType=map
+	// +listMapKey=version
+	Versions []APIVersionDiscovery `json:"versions,omitempty" protobuf:"bytes,2,rep,name=versions"`
+}
+
+// APIVersionDiscovery holds a list of APIResourceDiscovery types that are served for a particular version within an API Group.
+type APIVersionDiscovery struct {
+	// Version is the name of the version within a group version.
+	Version string `json:"version" protobuf:"bytes,1,opt,name=version"`
+	// Resources is a list of APIResourceDiscovery objects for the corresponding group version.
+	// +listType=map
+	// +listMapKey=resource
+	Resources []APIResourceDiscovery `json:"resources,omitempty" protobuf:"bytes,2,rep,name=resources"`
+	// DiscoveryFreshness marks whether a group version's discovery document is up to date.
+	// "Current" indicates no problems when fetching the discovery document. "Stale" indicates
+	// that there was an error fetching the discovery document, and the current version may not
+	// be up to date.
+	Freshness DiscoveryFreshness `json:"freshness,omitempty" protobuf:"bytes,3,opt,name=freshness"`
+}
+
+// APIResourceDiscovery provides information about an API resource for discovery.
+type APIResourceDiscovery struct {
+	// Resource is the plural name of the resource.  This is used in the URL path and is the unique identifier
+	// for this resource across all versions in the API group.
+	// resources with non-"" groups are located at /apis/<APIGroupDiscovery.objectMeta.name>/<APIVersionDiscovery.version>/<APIResourceDiscovery.Resource>
+	// resource with "" groups are located at /api/v1/<APIResourceDiscovery.Resource>
+	Resource string `json:"resource" protobuf:"bytes,1,opt,name=resource"`
+	// ReturnType describes the type of serialization that will be returned from this endpoint.
+	ReturnType GroupVersionKind `json:"returnType" protobuf:"bytes,2,opt,name=returnType"`
```

Comment:
Saying "serialization" is wierd.  We generally say "the kind of object returned".  Two points here:

Have we used `ReturnType` in any other API, or is that a net new naming addition in this context?  If it's net new, I might lean towards the same language we use in our openapi schemas which is `ResponseSchema`, although schema in openapi means an openapi schema, so we might say `ResponseKind`

```
        "responses": {
          "200": {
            "description": "OK",
            "schema": {
              "$ref": "#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.APIResourceList"
            }
          },
          "401": {
            "description": "Unauthorized"
          }
        }
```
The endpoint does not have to return the given type, so I'd probably add an adjective (`typical` or `primary`) indicating the type is one of many, and a sentence indicating `APIs may return other object types at their discretion, such as error conditions, requests for alternate representations, or other operation specific behavior`.  Having this doc is a great opportunity to clarify our public API and make sure users aren't surprised (this is the future face of our APIs!).

Reply:
This is net new, renamed to ResponseKind and updated the description.